modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1106 @@
+  unsigned int n_node = vertices.size();
+
+  Point normal;
+  Point center;
+
+  findElemNormal(vertices, normal, center);
+
+  bool inside = false;
+  unsigned int counter = 0;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+  {
+    unsigned int iplus1 = (i < n_node - 1 ? i + 1 : 0);
+    Point middle2p = p - 0.5 * (vertices[i] + vertices[iplus1]);
+    const Point side_tang = vertices[iplus1] - vertices[i];
+    Point side_norm = side_tang.cross(normal);
+    Xfem::normalizePoint(middle2p);
+    Xfem::normalizePoint(side_norm);
+    if (middle2p * side_norm <= 0.0)
+      counter += 1;
+  }
+  if (counter == n_node)
+    inside = true;
+  return inside;
+}
+
+bool
+MeshCut3DUserObject::findElemNormal(const std::vector<Point> & vertices,
+                                    Point & normal,
+                                    Point & center) const
+{
+  unsigned int n_node = vertices.size();
+
+  center(0) = center(1) = center(2) = 0;
+  normal(0) = normal(1) = normal(2) = 0;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+    center += vertices[i];
+  center /= n_node;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+  {
+    unsigned int iplus1(i < n_node - 1 ? i + 1 : 0);
+    Point ray1 = vertices[i] - center;
+    Point ray2 = vertices[iplus1] - center;
+    normal += ray1.cross(ray2);
+  }
+  normal /= n_node;
+  Xfem::normalizePoint(normal);
+  return true; // to be done
+}
+
+void
+MeshCut3DUserObject::findBoundaryNodes()
+// this is a simple algorithm simply based on the added angle = 360 degrees
+// works fine for planar cutting surface
+// for curved cutting surface, need to re-work this subroutine to make it more general
+{
+  _boundary_map.clear();
+
+  unsigned int n_nodes = _cut_mesh->n_nodes();
+  Real angle[n_nodes];
+  unsigned int node_id[3];
+  const Real PI = 3.14159265358979323846264338327950288419716939937510582;
+
+  for (unsigned int i = 0; i < n_nodes; ++i)
+    angle[i] = 0;
+
+  const MeshBase::element_iterator elem_begin = _cut_mesh->elements_begin();
+  MeshBase::element_iterator elem_end = _cut_mesh->elements_end();
+  MeshBase::element_iterator elem_it = elem_begin;
+
+  for (; elem_it != elem_end; ++elem_it)
+  {
+    Elem * cut_elem = *elem_it;
+    double vertices[3][3];
+
+    for (unsigned int i = 0; i < 3; ++i)
+    {
+      Node * this_node = cut_elem->get_node(i);
+      Point & this_point = *this_node;
+      for (unsigned int j = 0; j < 3; ++j)
+        vertices[i][j] = this_point(j);
+      node_id[i] = this_node->id();
+    }
+
+    angle[node_id[0]] += Xfem::angle_rad_3d(&vertices[2][0], &vertices[0][0], &vertices[1][0]);
+    angle[node_id[1]] += Xfem::angle_rad_3d(&vertices[0][0], &vertices[1][0], &vertices[2][0]);
+    angle[node_id[2]] += Xfem::angle_rad_3d(&vertices[1][0], &vertices[2][0], &vertices[0][0]);
+  }
+
+  for (unsigned int i = 0; i < n_nodes; ++i)
+  {
+    if (abs(angle[i] - PI * 2) > PI * 2 * 0.001)
+    {
+      _boundary_map[i] = new CutPoint;
+      _boundary_map.at(i)->id = i;
```

Comment:
Don't use `at()`. It throws and may not quit cleanly on some clusters with MPI. In general, use `mooseAssert()` instead to make sure items exist before accessing.